llvm/lib/Transforms/Scalar/Scalarizer.cpp

```diff
@@ -699,6 +730,20 @@ bool ScalarizerVisitor::splitCall(CallInst &CI) {
   if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1))
     Tys.push_back(VS->SplitTy);
 
+  if (AreAllVectorsOfMatchingSize) {
+    for (unsigned I = 1; I < CallType->getNumContainedTypes(); I++) {
+      if (isVectorIntrinsicWithStructReturnOverloadAtField(ID, I)) {
+        std::optional<VectorSplit> CurrVS = getVectorSplit(
+            cast<FixedVectorType>(CallType->getContainedType(I)));
+        // This case does not seem to happen, but it is possible for
+        // VectorSplit.NumPacked >= NumElems. If that happens a VectorSplit
+        // is not returned and we will bailout of handling this call.
+        if (!CurrVS)
+          return false;
+        Tys.push_back(CurrVS->SplitTy);
+      }
```

Comment:
Skipping failure cases for `CurrVS` should happen outside of the if.  Before, this was under a branch checking if the type differed from the last field, so it was only necessary to check this when it changed, but now, we are gating this on whether this field type contributes to the overload type, so the check needs to be moved out of the if.
Also: we still don't bail if `CurrVS->NumPacked != VS->NumPacked`.

```suggestion
      std::optional<VectorSplit> CurrVS = getVectorSplit(
          cast<FixedVectorType>(CallType->getContainedType(I)));
      // When using ScalarizeMinBits, it is possible for
      // VectorSplit.NumPacked >= NumElems. If that happens a VectorSplit
      // is not returned and we will bailout of handling this call.
      // We also bailout if NumPacked does not match.
      if (!CurrVS || CurrVS->NumPacked != VS->NumPacked)
        return false;
      if (isVectorIntrinsicWithStructReturnOverloadAtField(ID, I))
        Tys.push_back(CurrVS->SplitTy);
```

Reply:
`CurrVS->NumPacked != VS->NumPacked`. doesn't seem right. Because  frexp first out param can be half, float, or double. the second out param is  an int 32. 

`NumPacked = ScalarizeMinBits / ElemTy->getScalarSizeInBits()`

By default, we won't notice this break because `ScalarizeMinBits ` is an option parameter set to 0. and its integer division. but if the option `ScalarizeMinBits`  is set high enough we will notice a difference in behavior.

Reply:
Oops, I resolved too quickly, as I thought you took the suggestion.

I still think `|| CurrVS->NumPacked != VS->NumPacked` is correct, for the same reason it's checked here:
https://github.com/llvm/llvm-project/blob/709abacdc350d63c61888607edb28ce272daa0a0/llvm/lib/Transforms/Scalar/Scalarizer.cpp#L710-L718

Reply:
Note: I had also modified the comment from `This case does not seem to happen, ...`, since it can happen when `ScalarizeMinBits` is used.